### PR TITLE
chore: `BraceTransformer` - don't touch curly index braces since 8.4, as it's not a valid syntax anymore

### PR DESCRIPTION
**Pull request by keradus**
good to merge if nothing will fail, perf optimization only
https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/8312#discussion_r1873369005

### CODE REVIEW

**Review by Wirone — DISMISSED** [review]

File: src/Tokenizer/Transformer/BraceTransformer.php [hunk 1]
```diff
@@ -195,6 +195,12 @@ private function transformIntoPropertyHookBraces(Tokens $tokens, int $index): vo
 
     private function transformIntoCurlyIndexBraces(Tokens $tokens, int $index): void
     {
+        // as $arr{index} is illegal since 8.4 - ParseError
+        // (and produces FatalError on 8.0-8.3)
+        if (\PHP_VERSION_ID >= 8_04_00) {
```
Comment (Wirone) [inline src/Tokenizer/Transformer/BraceTransformer.php:200]: Please help me understand why we skip transforming only for 8.4+ while this syntax does not work on 8.0+?

Comment (keradus) [reply src/Tokenizer/Transformer/BraceTransformer.php:200]: https://3v4l.org/69BjL

please help to make comment describe it better

Comment (Wirone) [reply src/Tokenizer/Transformer/BraceTransformer.php:200]: I get that it behaves differently but I asked about something different: why we allow transformer on PHP 8.0-8.3 to replace tokens to custom tokens that represent the [concept that is invalid](https://3v4l.org/5Bt6U#v8.0.0)? It does not make sense to me. `CT::T_ARRAY_INDEX_CURLY_BRACE_*` should be available in the tokens collection only for PHP <8.0.

Comment (Wirone) [reply src/Tokenizer/Transformer/BraceTransformer.php:200]: It's a clear bug to me and I believe we rather need to fix tests if something fails when the condition is based on PHP 8.0.

Comment (keradus) [reply src/Tokenizer/Transformer/BraceTransformer.php:200]: I believe this one is good as it reflects validation logic of built-in tokenizer.

if you want to extend it further to match `php -l` validation, I'm OKey with that. feel free to raise it as separated PR

File: src/Tokenizer/Transformer/BraceTransformer.php [hunk 2]
```diff
@@ -195,6 +195,12 @@ private function transformIntoPropertyHookBraces(Tokens $tokens, int $index): vo
 
     private function transformIntoCurlyIndexBraces(Tokens $tokens, int $index): void
     {
+        // as $arr{index} is illegal since 8.4 - ParseError
+        // (and produces FatalError on 8.0-8.3)
```
Comment (Wirone) [inline src/Tokenizer/Transformer/BraceTransformer.php:199]:
```suggestion
        // Support for fetching array index with braces syntax (`$arr{$index}`)
        // was deprecated in 7.4 and removed in 8.0. However, the PHP's behaviour
        // differs between 8.0-8.3 (fatal error in runtime) and 8.4 (parse error).
        // @TODO Do not replace `CT::T_ARRAY_INDEX_CURLY_BRACE_*` for 8.0-8.3
```

**Review by Wirone — CHANGES_REQUESTED** [review]

File: src/Tokenizer/Transformer/BraceTransformer.php [hunk 3]
```diff
@@ -195,6 +195,14 @@ private function transformIntoPropertyHookBraces(Tokens $tokens, int $index): vo
 
     private function transformIntoCurlyIndexBraces(Tokens $tokens, int $index): void
     {
+        // Support for fetching array index with braces syntax (`$arr{$index}`)
+        // was deprecated in 7.4 and removed in 8.0. However, the PHP's behaviour
+        // differs between 8.0-8.3 (fatal error in runtime) and 8.4 (parse error).
+        // @TODO Do not replace `CT::T_ARRAY_INDEX_CURLY_BRACE_*` for 8.0-8.3
```
Comment (keradus) [inline src/Tokenizer/Transformer/BraceTransformer.php:201]:
```suggestion
        //
        // @TODO Do not replace `CT::T_ARRAY_INDEX_CURLY_BRACE_*` for 8.0-8.3, as further optimization
```

**Review by Wirone — APPROVED** [review]

**Pull request merged**